pages/design_documentation/ocf_io_engines.md

```diff
@@ -0,0 +1,269 @@
+blocks called IO Engines. These blocks implement specific algorithms, e.g.
+write-through or write-back and handle special (corner) cases as well.
+
+# OCF I/O engines - general description.
+OpenCAS can operate in [**several modes**](/cache_configuration.html), e.g.
+*write-back* or *write-through*. In general, OCF I/O engines are basic building
+blocks of a logic behind a particular cache mode. In other words, engines are
+sets of functions implementing the whole logic behind write and read operations
+for a given cache mode (and more, see below).
+
+This is an excerpt of a mapping table used in the OCF to call read or write
+handler of a particular engine (*write-through* and *write-back* engines 
+mappings are shown).
+
+~~~~~~~
+static const struct ocf_io_if IO_IFS[OCF_IO_PRIV_MAX_IF] = {
+	[OCF_IO_WT_IF] = {
+		.read = ocf_read_generic,       
+                .write = ocf_write_wt,          
+                .name = "Write Through"
+	}
+        [OCF_IO_WB_IF] = { 
+                .read = ocf_read_generic,       
+                .write = ocf_write_wb,          
+                .name = "Write Back"            
+        },
+[...]
+~~~~~~~
+
+For these particular engines, only the write path logic is different and
+specific engine functions are assigned accordingly. The read path is handled 
+by a generic read function.
+
+## I/O request handling by the engines
+
+When an I/O request is issued from an application using the OpenCAS software, 
+it is passed from the exported device (usually named /dev/casX-X) into the 
+OpenCAS Framework. The framework then determines which engine to use to handle 
+this request, and then a proper handling function is called (e.g., for a write 
+request to a cache running in a *write-back* mode - *ocf_write_wb()* will 
+be called).
+
+Most of the time, the engine implementing a configured cache mode is used. 
+However, in some less usual cases, a different I/O engine may be used 
+regardless of the configured cache mode (e.g. sequential cutoff threshold hit,
+cache drive I/O error, unaligned I/O, etc.).
+
+
+# OCF I/O engines - detailed description.
+Open CAS Framework uses the following engines:
+
+* [**Write Through**](/ocf_io_engines.html#write-through)
+* [**Write Back**](/ocf_io_engines.html#write-back)
+* [**Write Around**](/ocf_io_engines.html#write-around)
+* [**Pass Through**](/ocf_io_engines.html#pass-through)
+* [**Write Invalidate**](/ocf_io_engines.html#write-invalidate)
+* [**Write Only**](/ocf_io_engines.html#write-only)
+
+OCF also uses these internal engines that are not directly accessible through
+the OCF API, but are used in specific, unusual situations:
+
+* Fast
+* Discard
+* Direct To Core
+* Ops Engine
+
+
+## Write Through
+This engine is used to accelerate read operations, as the data is written to 
+both cache and core devices simultaneously.
+See [**write through cache mode configuration**](/cache_configuration.html#write-through)
+for details.
```

Comment:
This description confuses engine_wt with cache mode WT. Acceleration of reads is a consequence of combining engine_wt with engine_rd, not intrinsic property of engine_wt. What engine_wt actually does is what its name says - it writes through the cache down to the core volume. The major feature here is insertion of clean cache lines during the write operation.